docs/design/proposals/stateless-csi-plugin.md

```diff
@@ -0,0 +1,507 @@
+## Proposed solution details
+
+### Summary
+
+- Create a config file for the CSI plugins mapped to `/etc/ceph-csi-config/config.json`
+  - **NOTE:** This would get mapped into the plugin container as a config map
+  in kubernetes CO environments
+- Details about the Ceph cluster are maintained in `config.json`. Details
+  include a cluster-id, and MONs. These are updated as new clusters are
+  used to provision volumes from, or as MON addresses are changed for
+  existing clusters
+- VolumeCreate requests come against a targeted Ceph cluster-id and pool or fsname,
+  enabling the CSI plugin to locate the appropriate configuration information
+  for the cluster to act on the request
+  - **NOTE:** The cluster-id and pool/fsname information is specified in the
+  StorageClass for example
+- The returned *volume_id* from VolumeCreate requests, carries information
+  regarding the *cluster_id* and encoded pool/fsname IDs, enabling *volume_id* based
+  RPC requests to operate using the provided configuration information
+  against the respective cluster
+- Usual CSI spec based parameter maps such as, *volume_context* and *publish_context*
+  continue to carry information regarding required aspects about the
+  provisioned volume to dependent RPCs
+- The NS also has access to the `config.json` data to satisfy required node
+  level RPC requests
+
+### Pros
+
+- Addresses removal of the current config map that maintains list of
+  provisioned volumes
+- Provides the ability to have a single Ceph-CSI plugin instance per CO
+  instance, even when the CO instance needs volumes provisioned from multiple
+  Ceph clusters
+- Also, does not take away the capability, where needed, to use multiple
+  Ceph-CSI plugin instances (specialized using provisioner name) per CO
+  instance
+
+### Cons
+
+- Only RPCs that support passing credentials via secrets to the plugin
+  would work
+  - Some of the RPCs as of v1 of the CSI specification that would hence
+  not be supported are,
+    - Listing volumes and snapshots
+    - Getting capacity of the backing Ceph cluster(s)
+
+### Configuration specifics
+
+#### `config.json` config map contents
+
+```json
+{
+  "<clusterID>": {
+    "monitors": [
+      "IP/DNS:port",
+      ...
+    ]
+  },
+  ...
+}
+```
+
+#### Annotated YAML for RBD StorageClass
+
+```yaml
+kind: StorageClass
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: fast
+provisioner: csi-rbd
+parameters:
+  # [Optional] defaults to "ext4" allowed values depend on filesystems
+  # supported
+  fsType: ext4
+  # [Optional] defaults to "2"
+  imageFormat: "2"
+  # [Optional], can be specified when imageFormat is used, defaults to "",
+  #     possible values "layering"
+  imageFeatures: "layering"
+  # [Mandatory] carries the fsid of the Ceph cluster
+  clusterID: <ID>
+  # [Mandatory] carries the pool name to use in the Ceph cluster defined
+  #     by clusterID
+  pool: <pool-name>
+  # [Mandatory] All of the following are optional, and as described in,
```

Comment:
All of the following are optional or mandatory?

Reply:
The following are mandatory, need to change the text in both places.